### DISCUSSION TITLE
How to use path aliases within a packages directory?

### DISCUSSION THREAD

**Discussion by himantyu-yuma** [vercel/turborepo#4304] [Help] [answered]
I am currently developing an application using turborepo and Remix, and I would like to use path aliases within the packages directory.
For example, if I have following directory:
```
.
└── packages/
    └── ui/
        ├── Button/
        │   └── button.tsx
        └── Dialog/
            └── dialog.tsx
```
I want to import `Button` in `Dialog` using this syntax: `import { Button } from "@/Button"`.

I tried setting up the `packages/ui/tsconfig.json` file like this:
```json
{ 
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
      "@/*": ["./*"], 
    },
  }
} 
```
But I got errors and the path didn't resolve.

I read this Discussion( #620  ), but I want to use `@` as root of package.

If anyone knows how to set up path aliases in a monorepo setup using turborepo for importing packages.
Also, please let me know if this isn't a common practice when using monorepo setup or if there are better approaches. I'm not very familiar with monorepos.

**Comment by anthonyshew** [comment, marked as answer]
This is a perfectly fine practice! But it does have another layer of configuration to make work. 😄 

- [Subpath imports](https://nodejs.org/api/packages.html#subpath-imports) tell Node.js compilers where to look.
- `paths` in `tsconfig` tell TypeScript where to look.

What you end up with is a mismatch between what your compiler sees and what you see in your editor.

So, to use subpath imports, you'll need to follow the rules of Node.js (since they're more strict than TypeScript here.

In `packages/ui/package.json`:
```
  "imports": {
    "#*": "./src/*"
  },
```

And in `packages/ui/tsconfig.json`:
```
  "compilerOptions": {
    "paths": {
      "#/*": ["./src/*"]
    }
  },
```

Now, everyone speaks the same importing language and you can build successfully. 🥳

**Reply by singh-jay** [reply]
I have changed the imports and paths as suggested by @anthonyshew, still getting the module not-found error.
Module not found: Request should not start with "#/"
  1 | import { InputHTMLAttributes } from "react";
> 2 | import { BaseInput } from "#/primitive/base-input";

I tried different approaches based on the different threads but couldn't get it working.

Here is a demo repo - https://github.com/singh-jay/turbo-path-alias

Please share any reference repo, if anyone has a working solution.

**Reply by anthonyshew** [reply]
TypeScript now has automatic support for `package.json#paths` so you can get rid of the configuration in the `tsconfig.json`.

Note: You won't be able to use `#/*`. You'll need to use `#*` instead.

https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#auto-import-support-for-subpath-imports

**Reply by iantanwx** [reply]
@singh-jay you also need to ensure that the extensions are appropriately specified.

```
  "imports": {
    "#*": "./src/*",
    "#icons": "./src/icons.tsx",
    "#lib/*": "./src/lib/*.ts",
    "#atoms/*": "./src/atoms/*.tsx",
    "#styles/*": "./src/styles/*.css"
  }
```
 this is what my package.json looks like. hope that helps.

**Reply by braden-w** [reply]
Just a quick summary for all who may have stumbled into the above:

A modern monorepo-friendly alternative to Typescript path aliases is Node.js subpath imports, as recommended in [this GitHub issue](https://github.com/vercel/turborepo/discussions/620) and [the official Turborepo docs](https://turbo.build/repo/docs/guides/tools/typescript#use-nodejs-subpath-imports-instead-of-typescript-compiler-paths). Usage is almost identical to a Typescript path alias, except that they are defined in your `package.json` and must start with a `#` (this is a [strict requirement by Node.js](https://nodejs.org/api/packages.html#subpath-imports) to ensure these imports are clearly distinguished from external package specifiers)

```json
{
  "imports": {
    "#*": "./src/*.ts",
    "#db/schema": "./src/db/schema/index.ts"
  },
}
```

Adjust all of your imports accordingly:

```typescript
// Old style with relative paths
import type { Filter } from "../../conditions"
import { pagesTable } from "../db/schema";

// New style with subpath imports
import { Filter } from "#conditions";
import { pagesTable } from "#db/schema";
```

**Reply by karatashsyn** [reply]
> @singh-jay you also need to ensure that the extensions are appropriately specified.
> 
> ```
>   "imports": {
>     "#*": "./src/*",
>     "#icons": "./src/icons.tsx",
>     "#lib/*": "./src/lib/*.ts",
>     "#atoms/*": "./src/atoms/*.tsx",
>     "#styles/*": "./src/styles/*.css"
>   }
> ```
> 
> this is what my package.json looks like. hope that helps.

What if we have different types of files under a folder

**Reply by Nikola-Milovic** [reply]
I have the same question as @karatashsyn , this all seems swell for a small project, but manually specifying every single directory/ file will turn quite cumbersome as the project grows. A solution that supports specifying `./src/**/*` or `#/*: ./src/*` would be much more robust

**Comment by drithh** [comment]
I fixed it, my typescript version 5.4.5

here's what i change

packages/ui/package.json
```typescript
"imports": {
    "#dep/*": [
      "./src/*.ts",
      "./src/*.tsx"
    ]
 },
```

and for packages/ui/tsconfig.json

```typescript
{
  "extends": "@repo/typescript-config/react-library.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "."
  },
  "exclude": ["node_modules", "dist"]
}
```

and this is how i import the component

```typescript
import { cn } from '#dep/lib/utils';
```

**Comment by issam-seghir** [comment]
Is there a better solution than using "#" and "imports"? This method loses autocomplete and triggers a "cannot find module" error, which is very disappointing. Therefore, using normal paths without aliases seems preferable.

**Reply by anthonyshew** [reply]
Using Node.js subpath imports are the correct solution. Auto-complete works in latest versions of TypeScript and modules will resolve correctly in any LTS Node.js version.

It sounds like there may be a misconfiguration in your setup.

**Reply by Hyperion147** [reply]
Did you fix it @issam-seghir ?

**Reply by issam-seghir** [reply]
> Did you fix it @issam-seghir ?

No, I just stopped using Turborepo to avoid these headaches

**Reply by Nikola-Milovic** [reply]
@issam-seghir How does turborepo cause this issue? What alternative solution did you choose?
